execution-plane/arcon/src/streaming/source/kafka.rs

```diff
@@ -1,41 +1,281 @@
+            match message {
+                Ok(Ok(m)) => messages.push(m.detach()),
+                Ok(Err(KafkaError::NoMessageReceived)) => break, // No more messages pending
+                Err(_) => error!(self.ctx.log(), "Error while reading from stream."),
+                Ok(Err(e)) => error!(self.ctx.log(), "Kafka error: {}", e),
+            };
+            counter = counter + 1;
+            if counter == self.batch_size {
+                break;
+            }
+        }
+        // Process the batch
+        for m in messages {
+            let payload = match m.payload_view::<str>() {
+                None => "",
+                Some(Ok(s)) => s,
+                Some(Err(e)) => {
+                    error!(self.ctx.log(), "Error while deserializing message payload: {:?}", e);
+                    ""
+                },
+            }; /*
+            let key = match m.key_view::<str>() {
+                None => "",
+                Some(Ok(s)) => s,
+                Some(Err(e)) => {
+                    error!(self.ctx.log(), "Error while deserializing message key: {:?}", e);
+                    ""
+                },
+            }; */
+            let mut timestamp: Option<u64> = None;
+            if let Some(ts) = m.timestamp().to_millis() {
+                timestamp = Some(ts as u64);
+                if ts as u64 > self.max_timestamp {
+                    self.max_timestamp = ts as u64;
+                }
+            }
+            if let Ok(data) = serde_json::from_str(&payload) {
+                self.output_event(data, timestamp);
+            } else {
+                error!(self.ctx.log(), "Unable to deserialize message:\nkey: '{:?}', payload: '{}', topic: {}, partition: {}, offset: {}, timestamp: {:?}",
+                      m.key(), payload, m.topic(), m.partition(), m.offset(), m.timestamp());
+            }
+            /*
+            if let Some(headers) = m.headers() {
+                for i in 0..headers.count() {
+                    let header = headers.get(i).unwrap();
+                    println!("  Header {:#?}: {:?}", header.0, header.1);
+                }
+            }
+            // todo (somewhere else) 
+            consumer.commit_message(&m, CommitMode::Async).unwrap(); auto-committing
+            */
+        }
+        if counter > 0 {
+            self.output_watermark();
+            // Store Offset manage epoch here.
+        }
+        // Schedule next batch
+        self.schedule_once(Duration::from_millis(1000), move |self_c, _|{self_c.receive()});
+    }
+}
 
-*/
-#[derive(ComponentDefinition)]
-pub struct KafkaSource<A: 'static + ArconType> {
-    ctx: ComponentContext<KafkaSource<A>>,
-    target_pointer: ActorRef,
-    checkpoint_interval: u64,
-    _output_format: A,
+impl<OUT> Provide<ControlPort> for KafkaSource<OUT>
+where
+    OUT: 'static + ArconType + DeserializeOwned,
+{
+    fn handle(&mut self, event: ControlEvent) -> () {
+        match event {
```

Comment:
if let ControlEvent::Start = event {
   self.receive();
}

That should work. As the ControlEvent enum doesnt not have PartialEq, it doesn't work with a simple ==
or if

Reply:
Is there any reason to not use the match? I think it's a better pattern for the ControlEvent because we may want to add features using other ControlEvents.